nemo_skills/inference/prover.py

```diff
@@ -0,0 +1,428 @@
+                    and "failed" not in execution_result["stdout"]
+                ):
+                    results_dict["success"] = True
+                else:
+                    error_list = parse_error(execution_result["stdout"])
+                    error_message = get_error_str(full_code, error_list, error_thres=True)
+                    # checking for sorry
+                    if execution_result["process_status"] == "completed":
+                        stdout = execution_result["stdout"].lower()
+                        stderr = execution_result["stderr"].lower()
+                        combined = stdout + "\n" + stderr
+                        if re.search(r"\bsorry\b", combined) is not None:
+                            error_message += "\nThe code contains 'sorry', which means the proof is incomplete."
+                    if error_message.strip() == "":  # something in stderr indicating failure
+                        error_message = execution_result["stderr"][:1000]
+                        if len(execution_result["stderr"]) > 1000:
+                            error_message += "... (truncated)"
+
+                    last_error_message = (
+                        "We use <error></error> to signal the position of the error. \n" + error_message
+                    )
+                    feedback = self.refine_prompt.fill({"error_message": last_error_message})
+                    results_dict["feedback"] = feedback[0]["content"]
+                    results_dict["success"] = False
+
+            results_dict_list.append(results_dict)
+
+            if results_dict["success"]:
+                # This is the case when the code execution is successful. The theorem is proved.
+                break
+            else:
+                if self.cfg.refinement and turn_idx < self.cfg.refinement_max_turns - 1:
+                    prompt_turn_list += feedback
+                    full_prompt_turn_list += feedback
+                else:
+                    # Proving attempt failed.
+                    break
+
+        if len(results_dict_list) > 0 and results_dict_list[-1]["success"]:
+            success = True
+
+        # Usually only need prompt_turn_list for standard SFT, full_prompt_turn_list for SFT with remove_cot enabled, prompt_turn_list_list for SFT with delete_wrong_turns enabled.
+        return {
+            "code_list": code_list,
+            "results_dict_list": results_dict_list,
+            "prompt_turn_list": prompt_turn_list,
+            "turn_idx": turn_idx,
+            "success": success,
+            "full_prompt_turn_list": full_prompt_turn_list,
+            "prompt_turn_list_list": prompt_turn_list_list,
+        }
+
+    async def pass_at_N(self, data_point, data, N=None):
+        if N is None:
+            N = self.cfg.n_pass
+
+        new_results_dict = {"success": False}
+        for i in range(N):
+            results_dict = await self._single_data_point_generate(data_point, data)
```

Comment:
these are independent, but they are run serially. Use a loop like this to run them concurrently: https://github.com/NVIDIA-NeMo/Skills/blob/fcb687460a0d11349045fe6bc683ebbad24634f2/nemo_skills/inference/generate.py#L667-L674